### ISSUE TITLE
Transcode Rhai to S-Expressions

### ISSUE THREAD

**Issue by schungx** [lupyuen2/blockly-bl602#1]
Hi, maintainer of Rhai here.

I see that you're transcoding Rhai into an S-expression syntax.

Wonder how you're doing that and whether there is anything we can do to help.

**Comment by lupyuen** [comment]
Hi: Thank you so much for the awesome Rhai Scripting Engine :-)

I'm finishing my article that explains how I transcoded Rhai to uLisp S-Expressions...

-  ["Rust on RISC-V BL602: Rhai Scripting"](https://lupyuen.github.io/articles/rhai)

So far everything works great :-)

**Comment by schungx** [comment]
You're probably the first one (that I know of) who walks the AST...  is there anything that you find clumsy or hard to get around of?

I might also suggest:

```rust
const LED_GPIO = 11;

// instead of
let LED_GPIO = 11;
```

Using constants will propagate the number to all call sites, and you can get rid of one scope in uLisp because you can be sure the constant is not used anywhere in the script afterwards.  It is kept in the script because it creates a constant entry in the scope.

**Comment by lupyuen** [comment]
Thanks for the suggestion! I'm targeting Rhai Scripting for learners who are new to coding microcontrollers. So I'm pondering whether we should teach them `let` vs `const`, or maybe start with `let` and introduce `const` later? 

I'll publish the article to r/rhai real soon. Hope to generate some discussion over this :-)

**Comment by schungx** [comment]
I suppose in that case using `let` is one fewer keyword to confuse a new learner!

For a minimal language you might also want to disable keywords like `continue`, `break`, `switch`, `while`, `do` etc. and especially `eval` if you don't intend your users to touch them yet... since they would also need to be transcoded into uLisp...

I also opened an issue in your `iot` repo regarding compiling the Rhai engine for the BL602. I have seen Rhai compiled for the ST32 fitting under 64KB, so I wonder what is the issue you faced when using it for the BL602...  I suppose Rhai can't be _that_ much larger than uLisp...

**Comment by lupyuen** [comment]
Yep makes a lot of sense to reduce the keywords!

I might take some time and figure out why Rhai won't run on BL602. It's possible that the default FreeRTOS config on BL602 severely limits the Stack Space that I can use. (Sadly I'm no expert in FreeRTOS though)

**Comment by schungx** [comment]
I suppose you're building in release mode, right?  Rust's debug build uses a _LOT_ of stack space.  Stack space requirements are a lot lower for release builds.

And I see that you seem to be using a stack allocator, which may be why you run out of stack space so easily.  In that case, you'd want to minimize your data size; which means putting `no_position` as I suggested in that issue to save 32 bits from each AST node.

Finally, I'm not sure how large `Dynamic` is on 32-bit targets (which I assume the RISC-V is?)...  Let me check it out.  Maybe I'm using too large a tag data type.  EDIT: Just checked.  `Dynamic` is 8 bytes on 32-bit target with `no_float` and `only_i32` so that's OK.

EDIT 2: Don't forget LTO.  By default Rust turns OFF LTO, which matters a great deal to code size.  Also Rust defaults to optimize on speed; you might want size instead.

**Comment by lupyuen** [comment]
Yep I built it for release mode: [`run.sh`](https://github.com/lupyuen/bl_iot_sdk/blob/adc/customer_app/sdk_app_rust_script/run.sh#L30)

I'm using a Custom RISC-V Target (because the standard RV32 targets don't support hardware floating-point): [`riscv32imacf-unknown-none-elf.json`](https://github.com/lupyuen/bl_iot_sdk/blob/adc/customer_app/sdk_app_rust_script/riscv32imacf-unknown-none-elf.json)

Thanks for the tips, I'll try them as soon as I'm done with my article :-)

**Comment by schungx** [comment]
Remember to explicitly turn on LTO.  It usually makes a huge difference in code size, especially now that you're not using many of the functions.
